### ISSUE TITLE
Arguments to [DisallowNull] parameters sometimes aren't updated to NotNull state

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#50602]
**Version Used**: 
```
Branch master (14 Jan 2021)
Latest commit f3d7c35 by Julien Couvreur:
Don't check initializer on static record ctor (#50176)
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C1<T>
{
  void M1(T t)
  {
    Test(t);
    Test(t); // no warnings
  }
  
  void M2([AllowNull] T t)
  {
    Test(t);
    Test(t); // false CS8607
  }
  
  public void Test([DisallowNull] T s) { }
}
```

**Expected Behavior**:
No warnings for the second `Test(t)` call in both `M1` and `M2`.

**Actual Behavior**:
False `CS8607: A possible null value may not be used for a type marked with [NotNull] or [DisallowNull]` is reported for the second `Test(t)` call in `M2`

**Notes**
Note that in both cases `t` is passed to an argument disallowing null values and the state of the variable should be updated to non-nullable. However it seems that the compiler only updates `MaybeNull` state and leaves `MaybeDefault` as is.

**Comment by RikkiGibson** [comment]
To be clear, it's expected that the first invocation of `Test` in each method body gives a warning, but not the second, right?
